protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java

```diff
@@ -102,12 +104,33 @@ public static void initializeHAProxySupport(Session session, Channel channel) {
         channel.pipeline().addLast("proxy-protocol-packet-sender", new ChannelInboundHandlerAdapter() {
             @Override
             public void channelActive(ChannelHandlerContext ctx) throws Exception {
-                InetSocketAddress remoteAddress = (InetSocketAddress) ctx.channel().remoteAddress();
-                HAProxyProxiedProtocol proxiedProtocol = clientAddress.getAddress() instanceof Inet4Address ? HAProxyProxiedProtocol.TCP4 : HAProxyProxiedProtocol.TCP6;
+                HAProxyProxiedProtocol proxiedProtocol = getProxiedProtocol(clientAddress);
+
+                SocketAddress remoteAddress = ctx.channel().remoteAddress();
+                String destinationAddress;
+                int destinationPort;
+                if (remoteAddress instanceof InetSocketAddress inetRemoteAddress && getProxiedProtocol(inetRemoteAddress) == proxiedProtocol) {
+                    destinationAddress = inetRemoteAddress.getAddress().getHostAddress();
+                    destinationPort = inetRemoteAddress.getPort();
+                } else {
+                    // Fill in arbitrary values for the destination address and port if the remote address is not of the same type
+                    switch (proxiedProtocol) {
+                        case TCP4 -> {
+                            destinationAddress = "0.0.0.0";
+                            destinationPort = 0;
+                        }
+                        case TCP6 -> {
+                            destinationAddress = "0:0:0:0:0:0:0:0";
+                            destinationPort = 0;
+                        }
+                        default -> throw new UnsupportedOperationException("Unsupported proxied protocol: " + proxiedProtocol);
+                    };
```

Comment:
just in case this ever causes issues, let's a debug message stating what the destination address has been set to?